### ISSUE TITLE
Respond to second review

### ISSUE THREAD

**Issue by Robinlovelace** [ITSLeeds/stats19#63]
Thanks for the detailed review.
We think all the suggestions make sense, and link to previous discussions about combining the 3 stage (`dl`, `read`, `format`) process into a single function call, which we were thinking of calling `get_stats19()`.
There reason for splitting the process up is to ensure maximum transparency and to give the user control over what the package is doing.
However, as long as it is properly documents, we think the benefits of a `get_stats19()` function will outweigh any possible negatives we can think of so we plan to go ahead and create this function.

We would very much welcome a pull request that address some of the other issues you mention.

Responses to other issues/questions/comments are provided below.
We suspect that some of the comments refer to an older version of the package, which is completely understandable (the package has evolved since the initial submission!) and explains why some of the responses are short / questions.

- [x] Clearly much effort has gone into this package. I greatly support the sentiment behind making these data available in R having done the same for other data sets, myself. This package should be a great benefit to researchers in this area. I really appreciate the slim dependencies. It will make this package much easier to maintain into the future.

Thanks for the comments, we have indeed tried to keep dependencies to a minimum but consider `readr` and `tibble` worthwhile.
`readxl` and `curl` have been demoted to `Suggests`, as detailed in another comment.

- [x] I found the package to be well documented, the vignette is helpful in illustrating some use cases for the data along with how to access it and the code is clear.

Thanks. If you think of other was we can communicate the value of the data, do let us know (I think the second mapping figure could be improved...).

- [x] Some of the functionality I find to be mildly confusing like downloading and then importing the files into the R session and then formatting. As a user I'd prefer it all in one step, but there are likely use cases I'm not aware of that mean that this is useful so some examples of this would be useful I think.

We have long been planning to add a `get_stats19()` function as per https://github.com/ITSLeeds/stats19/issues/11 
The review comment, combined with further discussion, has triggered us to re-prioritise it.
It's been beneficial to polish each of the component functions first, however, and good to document each stage for maximum transparency, however, so we plan to keep the `dl`, `read` and `format` functions exported.

- [x] My general comments on the code follow and sections for the DESCRTIPTION and Vignette as well. I've commented quite a bit on grammar and spelling as I think that the polish of a package is important as it lends to the perception of the quality.

Agreed.

--------------------------------------------------------------------------------

- [x] Per rOpenSci policy, avoid start-up messages, rely on the documentation for
citation information:
https://ropensci.github.io/dev_guide/building.html#recommended-scaffolding.

The guidance is to 'Only use package startup messages when necessary'.
A case can be made that this is necessary.
As with `osmdata`, the package provides access to data that has a license that requires it to be cited.
The `osmdata` load message is as follows:

```{r}
library(osmdata)
```

We fully agree with the reasoning behind remove package startup messages however.
As a compromise, we've shortened the startup from 4 lines to 2:

```{r}
# before:
# Data provided under the conditions of the Open Government License.
# If you use data from this package, mention the source
# (UK Department for Transport), cite the package, and link to:
# www.nationalarchives.gov.uk/doc/open-government-licence/version/3/.
```

```{r}
# after:
library(stats19)
```

- [x] Avoid long lines >80 chars

running `goodpractice::gp()` found the following lines with > 80 lines:

```
    R/format.R:62:1
    R/format.R:67:1
    R/read.R:141:1
    R/utils.R:167:1
```

All these have been fixed.

- [x] Inconsistent use of white spaces in code, see `find_file_name()` `if` statements for examples.

- [x] The package does not pass `R CMD check`. *curl*, *readxl* and *tibble* are all listed as an Imports in DESCRIPTION but not imported from. With *curl* being used in tests, this means it should be in Suggests, I think. The others should be removed.

Kindly make sure you [`build`](https://github.com/r-lib/gtable/issues/52#issuecomment-181416744) before running `check the resulting `gz` archive.

`curl` is used in the tests and `readxl` is used in the examples.
These have been demoted to `Suggests`.
`tibble` has been removed from the DESCRIPTION file.

- [x] I don't think it's good form to have an example that won't work on Windows in the help file for `stats19_schema`, from data.R - line 17? Most of what I see there would be better served in a `data_raw` folder showing how the data were created with the documentation actually documenting what the variables are not how they were created, see <http://r-pkgs.had.co.nz/data.html> and for an example, <https://github.com/ropensci/GSODR/tree/master/data-raw>.

See https://github.com/ITSLeeds/stats19/pull/68

- [x] I would suggest to use proper formatting in help files, when naming packages, e.g. \pkg{stats19} and when referring to documented functions or data, e.g. \code{\link{stats19_schema}}, or with single quotes around abbreviations, e.g. 'DfT'. @ColinFay has an excellent page that outlines the formatting options and when/how to use them, <https://colinfay.me/writing-r-extensions/writing-r-documentation-files.html>. This will greatly enhance the users' experience when using the help files by making them more readable.

Acronyms have been explained and better formatting of 'STATS19' (note, it's capitalised when referring to the data, but lower case when referring to the package), and a few other improvements to the docs, have been made. See https://github.com/ITSLeeds/stats19/commit/e4ab09aa716630322aadd26dee3465258a2e58a0

- [x] I also would suggest making use of `@seealso` in documentation. For example, the `dl_stats19()` example works great in the help files, but from there I have the data but it's not in R. Using the `@seealso` you can let the user know about the `read_*()` functions.

`@seealso` tags have now been added to a few key documentation pages.

- [x] I downloaded files using `dl_stats19()`, selecting "Casualties", and then ran `read_accidents()` and got 
Is it possible to be more descriptive and say that I've used the wrong `read_*()` based on the file/data found and offer to import it?

Yes, but I think the costs of doing this would outweigh the benefits (RL)

- [x] Missing "." after "e.g." in dl.R on lines 8 and 9, there may be others that I didn't spy.

Fixed in https://github.com/ITSLeeds/stats19/commit/777479b9234be893a9946c787ef23cc08ff6a41a

- [x] Capitalisation in help files is inconsistent, e.g. lines 123-125 of read.R, parameter descriptions are mixed upper and lower case for first word after parameter itself. This applies to other functions where the descriptions are given in all lower case for other functions or upper case.

Fixed in https://github.com/ITSLeeds/stats19/commit/aec13fc0673717d68a3eb159c066d76a8d0645ca

- [x] Testing the functionality, I get this, when I expect it to tell me that `deaths` is not a valid input. But then when I hit escape, I expect it simply exit, not provide a warning message on the way out as well.

```r
dl_stats19(year = 1979, type = "deaths")
No files of that type found for that year.
This will download 240 MB+ (1.8 GB unzipped).
Files identified: Stats19-Data1979-2004.zip

Download now (y = enter, n = esc)? 

Warning message:
In find_file_name(years = year, type = type) :
  Coordinates unreliable in this data.
```

Now produces the following output: `Error in dl_stats19(year = 1979, type = "deaths") : Stopping as requested`

- [x] I got caught out when using the interactive features. I read "y = enter" but hit "y" thinking that would work as well as hitting "enter", but R cancelled the operation anyway just as if I'd hit "esc"

Agreed, that was an issue. Fixed in the latest version (tested with `> stats19::dl_stats19(year = 1979)`)

- [x] Per a recent conversation with CRAN, you should use `donttest()` rather than `dontrun()` for examples you don't want to be run on CRAN. Then set .travis.yml to run them by using `r_check_args: --as-cran --run-donttest`. **This may not be appropriate in all cases, e.g. interactive functions.**

Trying this, as per: https://github.com/ITSLeeds/stats19/pull/69

- [x] When validating user inputs and using `stop()` it's nice to use `call. = FALSE` to simplify the error message that the user receives.

Add as requested. See https://github.com/ITSLeeds/stats19/pull/70

- [x] Consider using [`hoardr`](https://ropensci.github.io/hoardr/) for managing user-saved files on disk that aren't in `tempdir()`?

We have considered this and have taken a good look at the package. In this case, however, I don't think it's appropriate: saving the 1/4 GB .zip file associated with the 1979:2005 data, for example, does not make sense when it can be extracted and saved in a more useful format. We assume the user will extract and save the data they use. Furthermore, the DfT is trialling an API which should make the desire to save intermediate files even rarer. Users are told where the files are saved, so are free to move them where they like.

- [x] When using `utils::download.file()`, you should use `mode = "wb"` or Windows users may end up with corrupted downloads in my experience. `curl::curl_download()` does the same thing but uses more updated ways of doing it and defaults to using a binary mode (wb).

We have tested on Windows and have not found any issues with the `dl_stats19()` function - so treating this as a 'wontfix', unless a strong reason can be found for adding this.

- [x] I don't think that there is much need for the `Attempting download from` or `Reading in: ` message. If it takes that long, I would suggest to use a progress bar to show progress. But this is just a personal observation.

There are read and download progress bars for large datasets. Try this, for example:

```r
dl_stats19(year = 1979)
```

- [x] Consider setting up a `pkgdown` site? It's easy to do and you can automate deployment with your Travis-CI so it's less to remember.

A site has been added: https://itsleeds.github.io/stats19/

I'm testing auto-deploy on travis and will happily merge this branch if it works (RL): https://github.com/ITSLeeds/stats19/pull/71

Update: auto deploy failed so reverting that change to .travis.yml. pkgdown that was making it fail. See https://travis-ci.org/ITSLeeds/stats19/builds/475415270#L1622 And: https://ropenscilabs.github.io/travis/

Is this is an ropensci pkg issue and is it worth asking for support?

#### Tests

- [x] I'm unclear how the interactive portion of the package functions is handled in testing? There are ways to handle this, but I don't see any implemented and when I run `devtools::test()` I'm asked to provide my own input.

When R is running non-interactively it auto-downloads the necessary files or produces an error. All tests involving downloads work on our computers and Travis. They can be turned off for the duration of your session with:

```r
Sys.setenv(DONT_DOWNLOAD_ANYTHING = "true")
```

Not sure if that answers the question - any further feedback on this: welcome.

- [x] Suggest using `skip_on_cran()` since some of the tests can take some time to execute due to download times.

Added in https://github.com/ITSLeeds/stats19/pull/72

#### DESCRIPTION File

See fixes here: https://github.com/ITSLeeds/stats19/commit/fdeb634dbab5a39ab599b35693623d0b1c94f71d

- [x] In the DESCRIPTION file, Mark's author entry is missing his ORCID.

- [x] More information in the DESCRIPTION's Description field would be desirable, a link to the data's website or other information to give more background perhaps.

- [x] STATS19 should be in "'" in DESCRIPTION for CRAN, i.e., 'STATS19', I think.

- [x] Check spelling in DESCRIPTION file, see: "analysie"

- [x] The Description should include a link to the DfT website.

- [x] Language field should be set, `Language: en-GB`

#### README File(s)

- [x] Use `remotes::install_github()` in place of `devtools::install_github()` in README.

- [x] The code style is inconsistent in the README.Rmd file in the code chunks, e.g. line 85 is missing space around `=`.

- [x] The example in the README showing two steps seems necessarily confusing to new users. If there is a good reason for having the raw data in R, document in a vignette why this is useful and show the two-step process, but if normal users won't do this, I wouldn't show it in the quick-start.

- [x] Line 43 of README uses inconsistent "(" around the phrases with the other two `read_*` function description.

#### Vignette

- [x] Run spell-check on it.

- [x] The term "attach"" has a specific meaning in R. Suggest rewording the portion about installation and loading the package to omit the use of "attach", since you're not using `attach()` in the R sense (and really shouldn't use it anyway).

- [x] I would describe why a user might want or need to install the Development version from GitHub in the vignette. Presumably if they are reading the vignette, they've already installed the package from CRAN (in the future).

- [x] Try to consistently use `function()` to identify functions in the vignette text. This also means that if/when you use pkgdown to build a site, the functions are linked to the help file.

- [x] In the introduction, the description of why there are `read_*()` and `format_*()` functions is confusing. To me, it reads as if `format` is only a parameter for `read_*()` in the introduction. I was left wondering why it's documented there or why the `format_*()`s even exist until I reached the end of the vignette.

- [x] There is a comma out of place in Vignette,

- [x] Format: Each of the read_*() functions has a format parameter which, when TRUE, adds

should be 

- [x] Format: Each of the read_*() functions has a format parameter, which, when TRUE, adds 

- [x] I'm unsure about including a package that's not on CRAN in the vignette (`ukboundaries`), something like this should be listed in Suggests, but it's not on CRAN, @sckott do you have any thoughts?

This is a good point.
Fixed, by adding a much more useful dataset, representing the juristictions of police forces across England and Wales: `police_boundaries`.

- [x] The first figures in the `sf` section after the join aren't immediately clear to me. The axis lack labels, I'm not really sure what I'm looking at.

Agreed. The maps were not particularly clear. An overview map has been added to the vignette. For context, and to ensure README-vignette harmony, we also changed the README's mapping section. It now also contains an overview map so people know what they are looking at, but not the facetted map, which was confusing.

#### Meta

- [x] The contributing guidelines mention a `pkgdown` website, this does not exist

**Comment by layik** [comment]
Great effort Robin, thank you.

**Comment by sckott** [comment]
regarding the `ukboundaries` not on CRAN thing, some have had luck with putting pkgs in a drat repo then adding the drat repository url to the DESCRIPTION file

**Comment by Robinlovelace** [comment]
Thanks for the supportive comment @sckott. Good news: the issue was fixed by adding a new dataset `police_boundaries` to the package (not in `ukboundaries`, yet...) that is actually much more relevant to the package. Will update the response accordingly!

**Comment by layik** [comment]
https://github.com/ropensci/software-review/issues/266

**Comment by layik** [comment]
I take pleasure of saying line 149 of read.R was 81 characters :)

**Comment by layik** [comment]
Now see if I can move the relevant bits to `raw` as per 3rd checkbox.

**Comment by layik** [comment]
Sending a PR soon. `download.file()` for which sorry?

**Comment by Robinlovelace** [comment]
For the file that contains the `police_boundaries` dataset. At the moment we take it directly from the internet.

**Comment by layik** [comment]
I see. Sure, can add commits to the branch. 
@Robinlovelace, could you please check the "spellcheck" commit, too? I think it is done but you know what it is like.

**Comment by Robinlovelace** [comment]
Will do. And push to master.
